src/pipeline/pipeline_construction.c

```diff
@@ -200,26 +239,55 @@ static ResultProcessor *getArrangeRP(Pipeline *pipeline, const AggregationPipeli
         ResultProcessor *rpLoader = RPLoader_New(params->common.sctx, params->common.reqflags, lk, loadKeys, array_len(loadKeys), forceLoad, outStateFlags);
         up = pushRP(&pipeline->qctx, rpLoader, up);
       }
-      rp = RPSorter_NewByFields(maxResults, sortkeys, nkeys, astp->sortAscMap);
-      up = pushRP(&pipeline->qctx, rp, up);
+
+      if (IsAggregate(&params->common) && HasWithCount(&params->common) && HasSortBy(&params->common)) {
+        size_t sort_limit = getSortLimit(params, astp, maxResults);
+        rp = RPSorter_NewByFields(sort_limit, sortkeys, nkeys, astp->sortAscMap);
+        up = pushRP(&pipeline->qctx, rp, up);
+      } else {
+        rp = RPSorter_NewByFields(maxResults, sortkeys, nkeys, astp->sortAscMap);
+        up = pushRP(&pipeline->qctx, rp, up);
+      }
     } else if (IsHybrid(&params->common) ||
-               IsSearch(&params->common) && !IsOptimized(&params->common) ||
+               (IsSearch(&params->common) && !IsOptimized(&params->common)) ||
                HasScorer(&params->common)) {
       // No sort? then it must be sort by score, which is the default.
       // In optimize mode, add sorter for queries with a scorer.
-      rp = RPSorter_NewByScore(maxResults);
+      if (IsAggregate(&params->common) && HasWithCount(&params->common) && HasSortBy(&params->common)) {
+        size_t sort_limit = getSortLimit(params, astp, maxResults);
+        rp = RPSorter_NewByScore(sort_limit);
+        up = pushRP(&pipeline->qctx, rp, up);
+      } else {
+        rp = RPSorter_NewByScore(maxResults);
+        up = pushRP(&pipeline->qctx, rp, up);
+      }
+    } else if (IsAggregate(&params->common) && HasDepleter(&params->common)) {
```

Comment:
This is inside `if (PipelineRequiresSorter(params))`
Might need to add a depleter also when the `if` condition is not met?

Reply:
Code added, but currently it's unreachable:
1. FT.AGGREGATE + WITHCOUNT disables optimization 
2. Without optimization `QOptimizer_QueryNodes()` is not called, and optimizer type is not modified.
3. For FT.AGGREGATE + WITHCOUNT we'll get the default value `Q_OPT_UNDECIDED`
4.So, `PipelineRequiresSorter(params)` will return always true because `params->common.optimizer->type != Q_OPT_NO_SORTER`